time-based-packet-transmission-offload.md

```diff
@@ -153,6 +171,29 @@ typedef enum _NDIS_TXTIME_SUPPORT_FLAGS {
 
 ```
 
+To obtain the `NDIS_TCP_IP_TXTIME_NET_BUFFER_LIST_INFO` structure, a driver should call the `NET_BUFFER_LIST_INFO` macro with an _Id of `TcpIpTxTimeNetBufferListInfo`
+```C
+typedef struct _NDIS_TCP_IP_TXTIME_NET_BUFFER_LIST_INFO {
+  struct {
+    ULONG IsIPv4 : 1;
+    ULONG IsIPv6 : 1;
+    ULONG IsTcp : 1;
+    ULONG IsUdp : 1;
+    ULONG TxTime : 64;
+    ULONG DropIfRate : 1;
```

Comment:
what is `DropIfRate`?

Reply:
Drop the packet if the packet arrived after deadline specified by TxTime. (Not sure here is the right place)

Reply:
We should not drop a packet for this reason. If it's later than the deadline, then we send it immediately. It's costly to recover this packet in TCP and it triggers congestion avoidance for no reason.

Reply:
Linux has the option though (I think, it means hardware has the feature already)

Reply:
For real-time applications where they don't care about loss, the drop might be ok? That being said, I kind of agree that this seems unnecessary and over complicates the design. Let's simplify things by removing support for this.

Reply:
Do you mind sharing a link? I am curious to learn why linux has it.

Reply:
As far as I undnerstand, the queue seems to be allocated on system/hardware. depends on offloading config (there is no explicit explanation though)